### PR TITLE
pytorch-0.3 with cuda and cudnn

### PR DESCRIPTION
**Pull request by akamaus**
###### Motivation for this change
Currently pytorch derivation lacks cuda support. After introducing cudatoolkit9 supporting gcc6 it became easy to add cuda-related flags.

###### Notes
Unfortunately cuda-enabled pytorch tests fail to detect driver if run by nix builder. Still they can be run manually after package installation. Just run unpackPhase and execute the script.
  
Had to provide a fixed commit fix hash for revision, without it getFetchFromGithub doesn't download required submodules.

### CODE REVIEW

File: pkgs/development/python-modules/pytorch/default.nix [hunk 1]
```diff
@@ -1,38 +1,52 @@
...
-    sha256 = "1s3f46ga1f4lfrcj3lpvvhgkdr1pi8i2hjd9xj9qiz3a9vh2sj4n";
+    rev    = "af3964a8725236c78ce969b827fdeee1c5c54110";
+    fetchSubmodules = true;
+    sha256 = "0zbndardq3fpvxfa9qamkms0x7kxla4657lccrpyymydn97n888a";
   };
 
-  checkPhase = ''
-    ${stdenv.shell} test/run_test.sh
-  '';
+  preConfigure = if cudnn != null then
```
Comment (FRidh) [inline pkgs/development/python-modules/pytorch/default.nix]: lib.optionalString

File: pkgs/development/python-modules/pytorch/default.nix [hunk 2]
```diff
@@ -1,38 +1,52 @@
...
 
   buildInputs = [
      cmake
      git
      numpy.blas
-  ];
+     utillinux
+     which
+  ] ++ lib.optionals cudaSupport
+        (lib.remove null [cudatoolkit cudnn]);
```
Comment (FRidh) [inline pkgs/development/python-modules/pytorch/default.nix]: no need to remove null

File: pkgs/top-level/python-packages.nix [hunk 3]
```diff
@@ -8658,7 +8658,19 @@ in {
...
   };
 
-  pytorch = callPackage ../development/python-modules/pytorch { };
+  pytorch = callPackage ../development/python-modules/pytorch {
+    cudaSupport = pkgs.config.cudaSupport or false;
+    cudatoolkit = pkgs.cudatoolkit;
+    cudnn = pkgs.cudnn;
+  };
+
+  pytorchWithCuda = self.pytorch.override {
```
Comment (FRidh) [inline pkgs/top-level/python-packages.nix]: why add extra attributes when its a matter of overriding and setting `cudaSupport=true;`?

**Review by FRidh — DISMISSED** [review]

**Comment by akamaus** [comment]
done

File: pkgs/development/python-modules/pytorch/default.nix [hunk 4]
```diff
@@ -1,38 +1,47 @@
...
     pyyaml
   ];
 
-  preConfigure = ''
-    export NO_CUDA=1
+  checkPhase = ''
+    ${stdenv.shell} test/run_test.sh
   '';
 
+  doCheck = false; # for some unknown reason doesn't detect cuda if run from builder user
```
Comment (orivej) [inline pkgs/development/python-modules/pytorch/default.nix]: Do the tests have to be disabled for `pytorchWithoutCuda` too?

Comment (akamaus) [reply pkgs/development/python-modules/pytorch/default.nix]: actually, not.

**Comment by tcosmo** [comment]
Hey, sorry I am a huge noob with NixOs, how can I use your fork in my nixpkgs so that I can finally run pytorch with cuda like calling something like :+1: `nix-env -iA nixos.python35Packages.pytorchWithCuda`
Thank you so much,

**Comment by akamaus** [comment]
@tcosmo, personally I just cherry-picked top two commits to channels/nixos-17.09. Now I can run pytorch like this:
> nix-shell  -p python3Packages.ipython  -p python3Packages.pytorchWithCuda

**Comment by FRidh** [comment]
cc maintainer @teh

**Comment by tcosmo** [comment]
@akamaus Thank you very much for your answer!!

File: pkgs/development/python-modules/pytorch/default.nix [hunk 5]
```diff
@@ -1,38 +1,47 @@
...
     pyyaml
   ];
 
-  preConfigure = ''
-    export NO_CUDA=1
+  checkPhase = ''
+    ${stdenv.shell} test/run_test.sh
   '';
 
+  doCheck = !cudaSupport; # for some unknown reason doesn't detect cuda if run from builder user
```
Comment (teh) [inline pkgs/development/python-modules/pytorch/default.nix]: Could the check failure be caused by group permissions? I think to use cuda you need to be in the video group? In any case disabling for the cuda case with comment is OK.

**Review by teh — APPROVED** [review]
LGTM. Thanks!

File: pkgs/development/python-modules/pytorch/default.nix (hunk 5 above)
Comment (akamaus) [reply pkgs/development/python-modules/pytorch/default.nix]: Well, I'm not sure. Users working with Cuda on my machine don't belong to video group. DISPLAY variable doesn't make any difference too.
Permissions for nvidia device are as follows:

> % ls /dev/nvidia0 -l                                                                                                                                 
> crw-rw-rw- 1 root root 195, 0 jan 14 18:35 /dev/nvidia0

**Comment by andersk** [comment]
This doesn’t build when merged to current master (or channels/nixos-17.09).

```
In file included from /nix/store/gv7w3c71jg627cpcff04yi6kwzpzjyap-cudatoolkit-9.1.85.1/include/host_config.h:50:0,
                 from /nix/store/gv7w3c71jg627cpcff04yi6kwzpzjyap-cudatoolkit-9.1.85.1/include/cuda_runtime.h:78,
                 from <command-line>:0:
/nix/store/gv7w3c71jg627cpcff04yi6kwzpzjyap-cudatoolkit-9.1.85.1/include/crt/host_config.h:121:2: error: #error -- unsupported GNU version! gcc versions later than 6 are not supported!
 #error -- unsupported GNU version! gcc versions later than 6 are not supported!
  ^~~~~
```

**Comment by teh** [comment]
Another issue: The non-cuda build seems to require cuda. I think it's down to this line:

```
preConfigure = lib.optionalString (cudnn != null) "export CUDNN_INCLUDE_DIR=${cudnn}/include";
```

**Comment by teh** [comment]
Another one: multiprocessig tests fail with ```RuntimeError: refcounted file mapping not supported on your system at /tmp/nix-build-python3.6-pytorch-0.3.0.drv-0/source/torch/lib/TH/THAllocator.c:525
```

I think the `optionalString` has to be sth like this:

```
preConfigure = lib.optionalString (cudaSupport && cudnn != null) "export CUDNN_INCLUDE_DIR=${cudnn}/include";```

@akamaus you still interested in updating this PR? Might be worth updating to 0.3.1.

**Comment by andersk** [comment]
I opened a new PR #38530 based on this, addressing all the above problems.

**Comment by FRidh** [comment]
Closing in favor of #38530.

**Pull request closed without merging**